restcomm/restcomm.telephony/src/main/java/org/restcomm/connect/telephony/CallManager.java

```diff
@@ -589,6 +616,71 @@ private void invite(final Object message) throws IOException, NumberParseExcepti
 
     }
 
+    /**
+     * FIXME: duplicated code make into static function or something more optimized
+     * Replace headers
+     * @param SipServletRequest message
+     * @param Map<String, ArrayList<String> > headers
+     */
+    private void addHeadersToMessage(SipServletRequest message, Map<String, ArrayList<String> > headers) {
+
+        if(headers!=null) {
+            for (Map.Entry<String, ArrayList<String>> entry : headers.entrySet()) {
+                //check if header exists
+                String headerName = entry.getKey();
+
+                StringBuilder sb = new StringBuilder();
+                if(entry.getValue() instanceof ArrayList){
+                    for(String pair : entry.getValue()){
+                        sb.append(";").append(pair);
+                    }
+                }
+                if(logger.isDebugEnabled()) {
+                    logger.debug("headerName="+headerName+" headerVal="+message.getHeader(headerName)+" concatValue="+sb.toString());
+                }
+                if(!headerName.equalsIgnoreCase("Request-URI")){
+                    try {
+                        String headerVal = message.getHeader(headerName);
+                        if(headerVal!=null && !headerVal.isEmpty()) {
+                            message.setHeader(headerName , headerVal+sb.toString());
+                        }else{
+                            message.addHeader(headerName , sb.toString());
+                        }
+                    } catch (IllegalArgumentException iae) {
+                        if(logger.isErrorEnabled()) {
+                            logger.error("Exception while setting message header: "+iae.getMessage());
+                        }
+                    }
+                }else{
+                    //handle Request-URI
+                    javax.servlet.sip.URI reqURI = message.getRequestURI();
+                    if(logger.isDebugEnabled()) {
+                        logger.debug("ReqURI="+reqURI.toString()+" msgReqURI="+message.getRequestURI());
+                    }
+                    for(String keyValPair :entry.getValue()){
+                        String parName = "";
+                        String parVal = "";
+                        int equalsPos = keyValPair.indexOf("=");
+                        parName = keyValPair.substring(0, equalsPos);
+                        parVal = keyValPair.substring(equalsPos+1);
+                        reqURI.setParameter(parName, parVal);
```

Comment:
@abdulazizali77 I believe we should completely replace the RURI instead of just modify the parameters
Can you point me to the design document for this as it seems I am not sync on the custom headers requirements